code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/DeploymentUnit.java

```diff
@@ -0,0 +1,149 @@
+package io.cattle.platform.servicediscovery.deployment.impl;
+
+import io.cattle.platform.core.model.Service;
+import io.cattle.platform.object.util.DataUtils;
+import io.cattle.platform.servicediscovery.deployment.DeploymentUnitInstance;
+import io.cattle.platform.servicediscovery.resource.ServiceDiscoveryConfigItem;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DeploymentUnit {
+
+    String uuid;
+    Map<Service, DeploymentUnitInstance> serviceToInstance = new HashMap<>();
+
+    public DeploymentUnit() {
+    }
+
+    public DeploymentUnit(String uuid, Map<Service, DeploymentUnitInstance> deploymentInstances) {
+        this.uuid = uuid;
+        serviceToInstance.putAll(deploymentInstances);
+    }
+
+    public boolean isError() {
+        /*
+         * This should check for instances with an error transitioning state
+         */
+        for (DeploymentUnitInstance instance : serviceToInstance.values()) {
+            if (instance.isError()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public void remove() {
+        /*
+         * Delete all instances. This should be non-blocking (don't wait)
+         */
+        for (DeploymentUnitInstance instance : serviceToInstance.values()) {
+            instance.remove();
+        }
+    }
+
+    public void stop() {
+        /*
+         * stops all instances. This should be non-blocking (don't wait)
+         */
+        for (DeploymentUnitInstance instance : serviceToInstance.values()) {
+            instance.stop();
+        }
+    }
+
+    public void start() {
+        /*
+         * Start the instances in the correct order depending on the volumes from.
+         * Attempt to start things in parallel, but if not possible (like volumes-from) then start each service
+         * sequentially.
+         * 
+         * If there are three services but only two containers, create the third
+         * 
+         * If one of the containers service health is bad, then create another one (but don't delete the existing).
+         * 
+         */
+        for (Service service : serviceToInstance.keySet()) {
+            createInstance(serviceToInstance, service, false);
+        }
+
+        for (DeploymentUnitInstance instance : serviceToInstance.values()) {
+            instance.waitForStart();
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    protected DeploymentUnitInstance createInstance(Map<Service, DeploymentUnitInstance> serviceToInstance,
+            Service service, boolean waitForStart) {
+        List<Integer> volumesFromInstanceIds = new ArrayList<>();
+        List<Integer> volumesFromServiceIds = DataUtils.getFields(service).get(
+                ServiceDiscoveryConfigItem.VOLUMESFROMSERVICE
+                .getRancherName()) == null ? new ArrayList<Integer>()
```

Comment:
nit: I'd rather not have cattle refer to `Rancher` by name.  Rancher is the product name, where as cattle is the name of the orchestration engine.  Cattle is an internal name that will probably never change, it's just a component.  Who knows tomorrow for marketing reason we might rename the company to Whale Wrangler.